Temp

uditagarwal97 opened the pull request:
Fio port for RTEMS

uditagarwal97 committed:
Added os-rtems.h to os

uditagarwal97 committed:
Delete os-rtems.h~

uditagarwal97 committed:
Push latest changes

uditagarwal97 committed:
Resolved few linker errors

uditagarwal97 committed:
Updated stat.c to remove redundant system usage variables

uditagarwal97 committed:
Included server.c and client.c and replaced legacy atomic functions

uditagarwal97 committed:
Removed redundant files

uditagarwal97 committed:
Removed final linker errors

uditagarwal97 committed:
Replaced user specific paths with env. variables

uditagarwal97 committed:
Update README

uditagarwal97 committed:
Updated makefile

uditagarwal97 committed:
Added media01 content to fio-rtems.c

uditagarwal97 committed:
Add automatic mounting of sd card before envoking RTEMS shell

uditagarwal97 committed:
Replaced MAP_SHARED with MAP_PRIVATE in mmap() for RTEMS with anonymous mappings

uditagarwal97 committed:
Set default pool size for RTEMS to 2

uditagarwal97 committed:
Added RTEMS namespace headers

uditagarwal97 committed:
Removed redundant files

uditagarwal97 committed:
fix invalid struct element error

uditagarwal97 committed:
Updated README

uditagarwal97 committed:
Added rtems shellfio command to fio.c

uditagarwal97 committed:
Resolved fatal shell exit and added remaining header files

uditagarwal97 committed:
fix option parsing. redefined constructors,destructors

uditagarwal97 committed:
Removed nice() usages in case of RTEMS

uditagarwal97 committed:
Removed headers

Headers were causing some additional tests to fail. Also, they won't be accepted by fio-master

uditagarwal97 committed:
Fixed minor spacing issues

uditagarwal97 committed:
Update RTEMS documentation, fix leftover spacing errors and remove fio-rtems.c

uditagarwal97 committed:
Merge pull request #1 from axboe/master

Merge latest changes

uditagarwal97 committed:
Minor trailing space fix and removed fio-rtems.c

uditagarwal97 committed:
few more space fixes, Arrh..

c-mauderer commented on Makefile:
```diff
@@ -194,11 +194,20 @@ ifneq (,$(findstring CYGWIN,$(CONFIG_TARGET_OS)))
+  FIO_OBJS +=  os/rtems/rtems-init.o
+endif
+
+OBJS := $(SOURCE:.c=.o)
```
I'm not sure if the `FIO_OBJS = $(OBJS) fio.o` still works as intended if you move the OBJS below that line. I would expect that it doesn't.

uditagarwal97 replied:
Actually, rtems-init.o has to be appended after fio.o. rtems-init.o contains dependencies from fio.o(included via fio-rtems-wrap.h)

c-mauderer replied:
OK. that explains why you added rtems-init.o as an object. But in the following snipped from your changed Makefile, `OBJS` is set after it is used. That seems not right.
```
FIO_OBJS = $(OBJS) fio.o

ifeq ($(CONFIG_TARGET_OS), RTEMS)
  FIO_OBJS +=  os/rtems/rtems-init.o
endif

OBJS := $(SOURCE:.c=.o)
```

c-mauderer commented on Makefile:
```diff
@@ -194,11 +194,20 @@ ifneq (,$(findstring CYGWIN,$(CONFIG_TARGET_OS)))
 FIO_OBJS = $(OBJS) fio.o
 
+ifeq ($(CONFIG_TARGET_OS), RTEMS)
+  FIO_OBJS +=  os/rtems/rtems-init.o
```
Wouldn't it be better to add the `os/rtems/rtems-init.c` to the SOURCE and not to the OBJS?

c-mauderer commented on README:
```diff
@@ -192,6 +192,25 @@ https://github.com/mintty/mintty/issues/56 and
+
+Toolchain, required for cross-compiling fio for desired architecture(like arm in this case) can be generated by using
+RTEMS Source Builder(RSB).
+Please note that fio build for RTEMS has been tested with RSB version 5, and it may not work with older versions.
```
Add the commit ID or the date of the versions you used for testing. Beneath that, add a note that you used RTEMS and rtems-libbsd.

c-mauderer commented on README:
```diff
@@ -192,6 +192,25 @@ https://github.com/mintty/mintty/issues/56 and
+
+Variable to be passed for toolchain path is TOOL_PATH_PREFIX, which in this case would be,
+
+uka_in@Madaari:~/development/benchmark/sandbox/5$ export TOOL_PATH_PREFIX=$PWD
```
I would suggest to use a plain `$` prompt instead of the `uka_in@Madaari:~/development/benchmark/sandbox/5$`. That one is used for other commands too. Beneath that, the README looks suspiciously like markdown or similar. Please use the same indentation like done for other commands (one space).

uditagarwal97 replied:
True, i'll fix that

c-mauderer commented on README:
```diff
@@ -192,6 +192,25 @@ https://github.com/mintty/mintty/issues/56 and
+
+make clean
+./configure --cc=../sandbox/5/bin/arm-rtems5-gcc --disable-optimizations --extra-cflags=-O3 --disable-pmem
+make CROSS_COMPILE=../sandbox/5/bin/arm-rtems5- V=1
```
Same as above: `$` as prompt and indentation by one space.

c-mauderer commented on cgroup.c:
```diff
@@ -224,7 +224,12 @@ void cgroup_shutdown(struct thread_data *td, struct cgroup_mnt *mnt)
+#else /* __rtems__ */
+static void fio_init
+#endif /* __rtems__ */
+cgroup_init(void)
```
Also that is a solution, maybe it would be a little shorter to re-define the `fio_init`. That would still mean that you need to change every line but it makes it a lot simpler to read.

I would suggest something like the following for `fio_init`:
```
#define fio_init(fname) static void __attribute__((constructor)) (fname)(void)
```
For RTEMS you make a second define without the attribute.
This gives you the possibility to use re-write the function headers as follows:
```
fio_init(cgroup_init)
{
   ...
}
```
Maybe we can use that macro sometimes to put the functions into some kind of linker set and call all functions during start. But that would be an extended goal for some far future.
I'm still not entirely happy with that but it is already slightly simpler to read. Most likely the FIO-developers will have another suggestion.
Same would be possible for fio_exit.

uditagarwal97 replied:
yes, that would be easier to comprehend. I'd prefer to present this solution to Fio's devs. If they agree with that, I'll then implement it

c-mauderer commented on crc/xxhash.h:
```diff
@@ -107,9 +107,16 @@ XXH32() :
 //****************************
 
+/* Unlike most desktop OS, In Newlib unsigned int and uint32_t are different */
+#ifndef __rtems__
```
Just a note: I think Sebastian fixed that recently. So _maybe_ with an up to date compiler / newlib, this isn't necessary any more.

uditagarwal97 replied:
Ok, i'll check this

uditagarwal97 replied:
~yes, this is now fixed~

c-mauderer commented on init.c:
```diff
@@ -7,9 +7,12 @@
+#ifndef __rtems__
 #include <dlfcn.h>
+#else /* __rtems__ */
+#include "dlfcn.h"
```
That's a local include. Where does it come from? Is that using the ./os/windows/posix/include/dlfcn.h?

uditagarwal97 replied:
No, it's from 5/arm-rtems5/beagleboneblack/lib/include/dlfcn.h, It appears after building rtems-libbsd with posix support

c-mauderer replied:
And it's in a local include path?

c-mauderer replied:
Just had a look: That should be found with <dlfcn.h> too. Otherwise there is some odd bug. What does your error message look like if you don't change that?

c-mauderer commented on lib/num2str.c:
```diff
@@ -1,3 +1,4 @@
+
```
Please don't add empty lines.

c-mauderer commented on options.c:
```diff
@@ -4782,6 +4782,11 @@ static char *bc_calc(char *str)
 	return strdup(buf);
+#else /* __rtems__ */
+	log_err("fio: performing math is not supported\n");
+	printf("fio: performing math is not supported\n");
```
A log and a print? Wouldn't one be enough?

c-mauderer commented on options.c:
```diff
@@ -4782,6 +4782,11 @@ static char *bc_calc(char *str)
+#else /* __rtems__ */
+	log_err("fio: performing math is not supported\n");
+	printf("fio: performing math is not supported\n");
+	return ' ';
```
That might isn't a good idea. It will return a pointer to a string where it isn't entirely clear were it comes from. That could theoretically be on the task stack. Either use a pointer to a `static const char[] = ' ';` or even better: return NULL like in other error cases. If there is math in the string, we can't do what the user want anyway. So an error is OK.

c-mauderer commented on os/os-rtems.h:
```diff
@@ -0,0 +1,71 @@
+/*-
+ * Copyright (c) 2018 Udit kumar agarwal <dev.madaari@gmail.com>
+ * Copyright (c) 2018 Christian Mauderer <christian.mauderer@embedded-brains.de>
```
It's nice that you mention me but I'm not aware that I did anything in that file. So I think you can just drop that line.

c-mauderer commented on os/os-rtems.h:
```diff
@@ -0,0 +1,71 @@
+
+static inline unsigned long long os_phys_mem(void)
+{
+    errno = ENOSYS;
```
You used some spaces for indentation here. Please use the same indentation style like everywhere else.

c-mauderer commented on os/os.h:
```diff
@@ -23,7 +23,7 @@ enum {
 	os_windows,
 	os_android,
 	os_dragonfly,
-
```
The space was most likely thought to separate the os_nr from the rest. So maybe put it below the new os_rtems.

c-mauderer commented on os/rtems/rtems-fio-wrap.h:
```diff
@@ -0,0 +1,54 @@
+#ifdef __rtems__
```
You already had that checked when you included that file. So you don't need it here.

c-mauderer commented on os/rtems/rtems-fio-wrap.h:
```diff
@@ -0,0 +1,54 @@
+	/* Destructors */
+	fio_syncio_unregister();
+    tiobench_unregister();
+    fio_filecreate_unregister();
```
Again: Mixed indentation style. Please fix the indentation for the whole file.

c-mauderer commented on os/rtems/rtems-init.c:
```diff
@@ -0,0 +1,242 @@
+int nice(int incr)
+{
+    /* FIXME */
+    return 0;
```
And again: Mixed indentation style.

c-mauderer commented on os/rtems/rtems-init.c:
```diff
@@ -0,0 +1,242 @@
+#define CONFIGURE_UNLIMITED_OBJECTS
+#define CONFIGURE_UNIFIED_WORK_AREAS
+
+#define CONFIGURE_STACK_CHECKER_ENABLED
```
For a performance test, you should disable the stack checker. Although it's quite handy during development, it can eat up a lot of performance.

c-mauderer commented on os/rtems/rtems-init.c:
```diff
@@ -0,0 +1,242 @@
+  &rtems_shell_IFCONFIG_Command, \
+  &rtems_shell_TCPDUMP_Command, \
+  &rtems_shell_SYSCTL_Command, \
+  &rtems_shell_VMSTAT_Command, \
```
I would suggest to remove most commands except the ones that you need for FIO. So you can shorten that to
```
#define CONFIGURE_SHELL_USER_COMMANDS &rtems_shell_fio_Command
```
and drop the rtems/netcmds-config.h.

Theoretically you could remove some from the list below too (everything that hasn't to do with file or disk handling. But Some diagnostic commands like CPUINFO might could be useful too.

c-mauderer commented on smalloc.c:
```diff
@@ -170,20 +174,24 @@ static bool add_pool(struct pool *pool, unsigned int alloc_size)
 	pool->map = ptr;
 	pool->bitmap = (unsigned int *)((char *) ptr + (pool->nr_blocks * SMALLOC_BPL));
 	memset(pool->bitmap, 0, bitmap_blocks * sizeof(unsigned int));
-
```
Why did you remove that line?

c-mauderer commented on workqueue.c:
```diff
@@ -136,13 +136,17 @@ static void *worker_thread(void *data)
 	}
-
+#else /* __rtems__ */
+	log_err("workqueue: nice %s\n", strerror(errno));
```
`errno` isn't set here like for the original case. Maybe replace the message with something other. Like `workqueue: no nice in RTEMS\n`.

uditagarwal97 committed:
Fix indentation, Remove redundant ifndef's

Dummy implementation of nice() has been done in rtems-init.c, so now there's no need of ifndef __rtems__ around nice in workqueue.c

uditagarwal97 committed:
Change dlfcn.h to <dlfcn.h>

Include path for dlfcn.h has been added to linker/compiler include paths

uditagarwal97 committed:
With recent changes in newlib, there's no need for subsituting uint32_t with unsigned int

uditagarwal97 committed:
Return NULL instead of ' ' in bc_calc for RTEMS, Fix Makefile

uditagarwal97 committed:
Remove redundant user shell commands

uditagarwal97 commented on crc/xxhash.h:
```diff
@@ -107,9 +107,9 @@ XXH32() :
 // Advanced Hash Functions
 //****************************
```
`crc/xxhash.c:263:15: error: conflicting types for 'XXH32_resetState'
 XXH_errorcode XXH32_resetState(void* state_in, uint32_t seed)
               ^~~~~~~~~~~~~~~~
In file included from crc/xxhash.c:65:0:
crc/xxhash.h:138:15: note: previous declaration of 'XXH32_resetState' was here
 XXH_errorcode XXH32_resetState(void* state, unsigned int seed);
               ^~~~~~~~~~~~~~~~
crc/xxhash.c:277:7: error: conflicting types for 'XXH32_init'
 void* XXH32_init (uint32_t seed)
       ^~~~~~~~~~
In file included from crc/xxhash.c:65:0:
crc/xxhash.h:110:15: note: previous declaration of 'XXH32_init' was here
 void*         XXH32_init   (unsigned int seed);` 
No, the issue is still there.

uditagarwal97 committed:
revert xxhash.h file to r7f01fba

uditagarwal97 committed:
Updated Documentation

uditagarwal97 committed:
Set INITIAL_POOL = 8 for RTEMS

Earlier there were some stack size related issues, due to which INITIAL_POOLS was set to 2 for RTEMS. There's no need for that now.

uditagarwal97 committed:
minor non-functional style fix
